### PR TITLE
feat(sdk.v2): Support optional component inputs in v2.

### PR DESCRIPTION
**Pull request by chensun**
**Description of your changes:**
IR doesn't allow optional component inputs at this moment. As a short term workaround, we add component inputs only when they were passed into the task instance. One drawback of this approach is that component spec is tied to a specific task -- reusing the same component multiple times will generate multiple component specs.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by chensun** [comment]
/hold

This PR contains commits from https://github.com/kubeflow/pipelines/pull/5382.
Review can wait until rebasing after https://github.com/kubeflow/pipelines/pull/5382 is merged.

**Comment by chensun** [comment]
/unhold

This is ready for review now.

**Comment by neuromage** [comment]
/lgtm

Let's open a bug to fix this in IR at some point down the road.

File: sdk/python/kfp/v2/compiler_cli_tests/test_data/pipeline_with_nested_conditions_yaml.json [hunk 1]
```diff
@@ -270,6 +270,36 @@
         },
         "executorLabel": "exec-print"
       },
+      "comp-print-2": {
```
Comment (Ark-kun) [inline sdk/python/kfp/v2/compiler_cli_tests/test_data/pipeline_with_nested_conditions_yaml.json:273]: Ohh...
Technically we could dedup these, but I think it's OK at this moment.
We need to add optional (and default) to MP and stop needing this workaround.

File: sdk/python/kfp/dsl/component_spec.py [hunk 2]
```diff
@@ -35,11 +35,14 @@ def additional_input_name_for_pipelineparam(
 
 def build_component_spec_from_structure(
     component_spec: structures.ComponentSpec,
+    actual_inputs: List[str],
 ) -> pipeline_spec_pb2.ComponentSpec:
   """Builds an IR ComponentSpec instance from structures.ComponentSpec.
 
   Args:
     component_spec: The structure component spec.
+    actual_inputs: The acutal arugments passed to the task. This is used as a
```
Comment (Ark-kun) [inline sdk/python/kfp/dsl/component_spec.py:44]: Typo: `actual`

Comment (chensun) [reply sdk/python/kfp/dsl/component_spec.py:44]: Fixed.

File: sdk/python/kfp/v2/compiler_cli_tests/test_data/pipeline_with_nested_conditions_yaml.json (hunk 1 above)
Comment (chensun) [reply sdk/python/kfp/v2/compiler_cli_tests/test_data/pipeline_with_nested_conditions_yaml.json:273]: Agree, and opened an internal feature request.

**Pull request merged**